tomcat/tomcat-jta/src/main/java/org/jboss/narayana/tomcat/jta/TransactionalDataSourceFactory.java

```diff
@@ -0,0 +1,183 @@
+import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
+
+import javax.naming.Context;
+import javax.naming.Name;
+import javax.naming.RefAddr;
+import javax.naming.Reference;
+import javax.naming.spi.ObjectFactory;
+import javax.sql.XADataSource;
+import javax.transaction.TransactionManager;
+import javax.transaction.xa.XAResource;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+import java.sql.SQLException;
+import java.util.Hashtable;
+import java.util.Properties;
+
+/**
+ * @author <a href="mailto:zfeng@redhat.com">Zheng Feng</a>
+ */
+public class TransactionalDataSourceFactory implements ObjectFactory {
+    private static final String PROP_USERNAME = "username";
+    private static final String PROP_PASSWORD = "password";
+    private static final String PROP_MAX_TOTAL = "maxTotal";
+    private static final String PROP_MAX_IDLE = "maxIdle";
+    private static final String PROP_MIN_IDLE = "minIdle";
+
+    private static final String[] ALL_PROPERTIES = {
+            PROP_USERNAME,
+            PROP_PASSWORD,
+            PROP_MAX_TOTAL,
+            PROP_MAX_IDLE,
+            PROP_MIN_IDLE
+    };
+
+    @Override
+    public Object getObjectInstance(Object obj, Name name, Context context, Hashtable<?, ?> environment) throws Exception {
+
+        if (obj == null || !(obj instanceof Reference)) {
+            return null;
+        }
+
+        final Reference ref = (Reference) obj;
+        if (!"javax.sql.DataSource".equals(ref.getClassName())) {
+            return null;
+        }
+
+        final Properties properties = new Properties();
+        for (final String propertyName : ALL_PROPERTIES) {
+            final RefAddr ra = ref.get(propertyName);
+            if (ra != null) {
+                properties.setProperty(propertyName, ra.getContent().toString());
+            }
+        }
+
+        TransactionManager transactionManager = (TransactionManager) getReferenceObject(ref, context, "transactionManager");
+        XADataSource xaDataSource = (XADataSource) getReferenceObject(ref, context, "xaDataSource");
+
+        XARecoveryModule xaRecoveryModule = getXARecoveryModule();
+        if (xaRecoveryModule != null) {
+            xaRecoveryModule.addXAResourceRecoveryHelper( new XAResourceRecoveryHelper() {
+                @Override
+                public boolean initialise(String p) throws Exception {
+                    return true;
+                }
+
+                @Override
+                public XAResource[] getXAResources() throws Exception {
+                    try {
+                        String user = properties.getProperty(PROP_USERNAME);
+                        String password = properties.getProperty(PROP_PASSWORD);
+
+                        if (user != null && password != null) {
+                            return new XAResource[]{xaDataSource.getXAConnection(user, password).getXAResource()};
```

Comment:
Good catch although the TransactionalDriver link you sent is about the general connection management rather than the specific recovery connection. I would propose that the recovery connection - the one we are talking about here - is cached for reuse. I don't think we have to close it though.

Reply:
Just to complete this thought: The equivalent of ConnectionListener as used in TransactionalDriver is expected to live within DBCP2 and should close the connection there

Reply:
I think the recovery connection is created in the same way as the `getXAResource` would be used (https://github.com/jbosstm/narayana/blob/master/ArjunaJTA/jdbc/classes/com/arjuna/ats/internal/jdbc/recovery/JDBCXARecovery.java#L117). But I'm pretty unsure here. Please let me know how is it.

Ah, agree the `ConnectionListener` is probably at the level of the dbcp2 just if there is not a chance that recovery would be trying to use a cached connection where it's underlaying counterpart is in error. Do I misinterpret here?

Reply:
I guess it depends on the database implementation? I would hope that most modern databases could recover from a transient connection problem and be re-used.

Reply:
Perhaps this is why @Karm runs out of connections?

Reply:
I created https://issues.jboss.org/browse/JBTM-3012